Bounce variables around forced stack allocations in RSSA loops

jasoncarr0 opened the pull request:
This is not currently ready to merge, but I'd like to put the code up here and have a thread for this. Given that oxygen is down, could you run the proper statistics on this? At the very least it should be tested on native/c/llvm, but may also be significantly affected by versions thereof.

I set the limit of variables per loop down to 2. This is quite weak, but it keeps the worst benchmark (wc-scanStream) from being slowed down as much as it was with more rewrites, although even a small number still has quite a strong effect. It's still enough to see improvements on some benchmarks.

DLXSimulator can be improved much more than this but I haven't been able to find any improved method for variable selection. It generally gets up to a limit of 32% improvement with large variable rewrite limits.

On C and LLVM codegens, penalties are usually much less and benefits persist. Some benchmarks and codegens (e.g. string-concat with the improved shrink, and peek with high quality llvm/gcc) manage to saturate their speedup without this, although this sometimes also contributes.

MatthewFluet commented:
Benchmark data (both execution time and compile time) are available at: https://www.cs.rit.edu/~mtf/mlton/reg-alloc/20190516/
These weren't run on oxygen, but on cadmium (a 48-core AMD machine, which has a slower clock speed than oxygen, but the many more cores allows the benchmarks to run in parallel).

jasoncarr0 commented:
I'm surprised about some points in this data. wc-scanStream apparently gets no slower, while I still see a 2x slowdown on my machine or oxygen, which does somewhat support the position that it's heavily machine dependent, and maybe should be ignored.
Knuth-bendix was another one which had slowdowns, but it's possible that it is actually getting a small speedup on cadmium.

So, maybe it is worthwhile bumping it back up to 8. A few benchmarks seem to really benefit from running it. I tried again running comparisons of various limits once oxygen was up, and it seems like so far 8 is indeed performing better in most cases, with the exception of course of wc-scanStream.

Overall I'm very happy to see the results. If it's not too difficult, perhaps we could also run these with `-bounce-rssa-limit 8` and confirm? I'm also reminded of one more restriction I haven't implemented to not consider GC points with many live variables.

jasoncarr0 commented:
Though, it is quite painful to see a slowdown on mlyacc, so perhaps there is yet more tuning.

MatthewFluet commented:
I added a `-bounce-rssa-limit 8` to the benchmarks.  (Since nothing else changed, I was able to extend the existing data.  But, note carefully the configuration number when comparing; the amd64 runs are C00, C01, and C06.)

`wc-scanstream` does now exhibit a larger slowdown with amd64.

jasoncarr0 commented:
> I added a `-bounce-rssa-limit 8` to the benchmarks. (Since nothing else changed, I was able to extend the existing data. But, note carefully the configuration number when comparing; the amd64 runs are C00, C01, and C06.)
> 
> `wc-scanstream` does now exhibit a larger slowdown with amd64.

Actually, it appears to be showing a speedup, as far as I can see: https://www.cs.rit.edu/~mtf/mlton/reg-alloc/20190516/report_programs.comp/report_wc-scanStream_task_clock.html#wc-scanStream::task_clock::ratios_data_summary

MatthewFluet commented:
`report_programs.comp` is compile-time; you want https://www.cs.rit.edu/~mtf/mlton/reg-alloc/20190516/report_programs.exec/report_wc-scanStream_task_clock.html#wc-scanStream::task_clock::ratios_data_summary

jasoncarr0 commented:
That helps make sense of things

jasoncarr0 commented:
So, it seems that cutting off bounce points based on bounce-able variables works pretty well. I've set the default limit to 12. I've also checked 16 and cannot notice any difference in the benchmarks. Right now, it seems a cutoff of 12 and limit of 12 (which means variables would be assigned from multiple bounce points or multiple loops) seems to be the best overall on native, although a more conservative 16 and 8 works well, but variable heurstics seem to still be picking poorly for e.g. DLXSimulator, which gets improved by the limit change 8 -> 12.

Here's data from AMD64 on oxygen:
[amd64-oxygen-1.txt](https://github.com/MLton/mlton/files/3202965/amd64-oxygen-1.txt)
MLton7/8 are the more conservative settings, while 2/3 have a higher limit. All of them avoid wc-scanStream problems, and 2 (With a cutoff of 8 and limit of 12) still has the speedup on DLXSimulator.

MatthewFluet commented:
I've update the benchmark results at https://www.cs.rit.edu/~mtf/mlton/reg-alloc/20190516/ with `-bounce-rssa-live-cutoff 12 -bounce-rssa-limit 8` and `-bounce-rssa-live-cutoff 12 -bounce-rssa-limit 12` for all three codegens.

MatthewFluet commented on mlton/backend/restore.sig:
```diff
@@ -0,0 +1,22 @@
+      include RSSA
+   end
+
+signature RESTORE_R =
```
I think that this should be named `RESTORE_RSSA`.  I know that we have `RESTORE` (for SSA) and `RESTORE2` (for SSA2), but this seems unnecessarily opaque.

MatthewFluet commented on mlton/control/control-flags.sml:
```diff
@@ -44,6 +44,19 @@ val atMLtons = control {name = "atMLtons",
+val bounceRssaLiveCutoff = control {name = "bounceRssaLiveCutoff",
+                               default = SOME 12,
+                               toString = Option.toString Int.toString}
+val bounceRssaLoopCutoff = control {name = "bounceRssaLiveCutoff",
```
```suggestion
val bounceRssaLoopCutoff = control {name = "bounceRssaLoopCutoff",
```

MatthewFluet commented on mlton/control/control-flags.sml:
```diff
@@ -44,6 +44,19 @@ val atMLtons = control {name = "atMLtons",
+val bounceRssaLoopCutoff = control {name = "bounceRssaLiveCutoff",
+                               default = SOME 40,
+                               toString = Option.toString Int.toString}
+val bounceRssaUsageCutoff = control {name = "bounceRssaLiveCutoff",
```
```suggestion
val bounceRssaUsageCutoff = control {name = "bounceRssaUsageCutoff",
```

MatthewFluet commented:
Here's a tweak to the `merge` benchmark to make the main computation tail recursive (and, therefore, implemented as loops):

``` sml
fun merge (l1: int list, l2) =
   let
      fun revapp (l, acc) =
         case l of
            [] => acc
          | h :: t => revapp (t, h :: acc)
      fun loop (l1, l2, acc) =
         case (l1, l2) of 
            ([], _) => revapp (acc, l2)
          | (_, []) => revapp (acc, l1)
          | (x1 :: l1', x2 :: l2') =>
               if x1 <= x2
                  then loop (l1', l2, x1 :: acc)
                  else loop (l1, l2', x2 :: acc)
   in
      loop (l1, l2, [])
   end

structure Main =
   struct
      fun doit size =
         let
            val len = 100000
            val l1 = List.tabulate (len, fn i => i * 2)
            val l2 = List.tabulate (len, fn i => i * 2 + 1)

            fun test () =
               if 0 = hd (merge (l1, l2))
                  then ()
               else raise Fail "bug"

            fun loop n =
               if n = 0
                  then ()
               else (test (); loop (n - 1))
         in
            loop size
         end
   end
```

This is the kind of program that I would expect to get a non-trivial speedup from moving variables from stack slots to locals.  However, I don't see significant differences between 05004a037, a4200bf0d, and 752090717.  In fact, if I'm reading the Machine IR correct, then with a4200bf0d and 752090717, the `x1` and `x2` variables in the merge loop are being stack allocated.  (One way to check is to tweak the benchmark to use `Int16.int list` and look for the one `WordS16_lt` primitive.)  Admittedly, the SSA/SSA2/RSSA/Machine IR code is a little more convoluted than one might expect.

jasoncarr0 commented:
I'm looking at that merge issue, and it seems like other variables are being bounced instead (for good reason). But there's also some weird points where bouncing is doing the right thing, but there are better solutions by just shrinking lifetimes: (this is pre-boucning)
```
  L_273 () Jump = 
    x_355: Objptr (opt_16) = OP (x_119, 8): Objptr (opt_16)
    x_354: Word16 = OW16 (x_119, 0): Word16
    L_114 (x_355, x_354, Cast (0x1:w64: Word64, Objptr (opt_18)))
  L_114 (x_173: Objptr (opt_16), x_172: Word16, x_171: Objptr (opt_18)) Jump =      
    x_396: Word32 = CPointer_lt (Limit, Frontier)
    switch {test = x_396,
            default = None,
            cases = ((0x0:w32, L_276), (0x1:w32, L_275))}
  L_275 () Jump =      
    CCall {args = (<GCState>, 0x0:w64: Word64, 0x0:w32: Word32),
           func = {args = (GCState, Word64, Word32),
                   convention = cdecl,
                   kind = Runtime {bytesNeeded = None,        
                                   ensuresBytesFree = Some 1,
                                   mayGC = true,
                                   maySwitchThreads = false,
                                   modifiesFrontier = true,
                                   readsStackTop = true,
                                   writesStackTop = true},        
                   prototype = {args = (CPointer, Word64, Int32), res = None},
                   return = Bits0,
                   symbolScope = private,
                   target = GC_collect},
           return = Some L_277}
  L_277 () CReturn {func = {args = (GCState, Word64, Word32),                                                                                                                                                                
                            convention = cdecl,
                            kind = Runtime {bytesNeeded = None,
                                            ensuresBytesFree = Some 1,
                                            mayGC = true,
                                            maySwitchThreads = false,
                                            modifiesFrontier = true,
                                            readsStackTop = true,
                                            writesStackTop = true},
                            prototype = {args = (CPointer, Word64, Int32),
                                         res = None},
                            return = Bits0,
                            symbolScope = private,
                            target = GC_collect}} =      
    L_276 ()
  L_276 () Jump =      
    x_356: Objptr (opt_18) = Object {header = 0x25, size = 24}
    x_397: Bits32 = WordU16_extdToWord32 (x_172)
    OW32 (x_356, 0): [Word16, Bits16] := Cast (x_397, [Word16, Bits16])
    OP (x_356, 8): Objptr (opt_18) := x_171
    switch {test = Cast (x_173, Bits64)
```

For example, follow x_354/x_172 here. It's created before the GC point and never used until afterwards, where the object it was derived from is still live. 

Because of usage counts, the bouncer prefers the objptr x_355 here, and bounces it instead. The best move is actually: eliminate the definition of x_172/x_354, since there are no uses before the GC, and since it is derived from x_355, which is live. Then, redefine x_354 after the GC point, before its first usage. This is notably somewhat similar to the idea for globals, simply copy them into a variable at loop entry.

This does seem a bit difficult to calculate, but is a good idea in general (check live ranges of objects, and variables derived from them, move the definition point to the last location before the first use, and within the live range of the base object). Perhaps more shrinking and copy propagation can help a tiny bit here, copy prop does not currently do anything with operands besides cast/const/var. The difficulty is still not inadvertently extending live ranges.

jasoncarr0 commented:
Regardless, it does feel like surprisingly few variables are bounced in general, so perhaps a different way to choose variables to bounce and weight them might be appropriate. The reasoning for counting by loop is that ranges will often overlap, and we'd like to avoid register pressure inside the loop proper, but in practice this means that many points will have one-two variables bounced, even though there are 5 more that get put in stack slots. So perhaps weighting variables  by cost and choosing by location is necessary.

MatthewFluet commented:
I think that what might be happening with `tailmerge` is that because many of the lists are "obviously" non-empty, they end up getting passes around as separate hd and tl vars, rather than as a single cons.  And that does seem to be artificially increasing the number of live variables at the GC point.

jasoncarr0 commented:
So, with regards to tailmerge, bouncing every variable seems to improve performance by under 10% (on native). Since this usually is a good indicator of the maximum (positive or negative), we wouldn't expect to see large gains with any particular choice of variables.

I've tried implementing the following: preferring variables with a smaller span over loops (I've reset it out of the branch, it's at 70d3ac8381e2bff68d311c3e7ab76ab2abc141d7), and weighting counts by the inverse of loop block count (i.e. if a loop is used twice in a loop of size 5, and once in a loop of size 10, it will have weight 2/5 + 1/10 = 1/2).

Neither of these do better than 100% on tailmerge, and may do slightly worse on some finicky programs (DLXSimulator etc). We could try to see if this or 70d3ac83 helps, but overall it looks about the same from what I can see.

jasoncarr0 commented:
This should be mostly good to go. I fixed an issue with cutoffs in b5702fd, (that I introduced in f9cd0dc0 and fixed incorrectly), so the performance characteristics may be slightly different.

MatthewFluet commented:
Sounds good.  I'm going to merge #308 into `master`, then merge `master` into this to resolve the merge conflicts.  Then I'll run the benchmarks to include the results in the merge commit.

jasoncarr0 commented:
So one last note. I haven't yet disabled the pass by default on LLVM, so we can look over the data with that, and verify if there's reason for it to be enabled or disabled (e.g. compile time or binaries even if execution speed is average 1.0).

Also, the pass was renamed to bounceVars. Control flags are still unchanged

MatthewFluet commented:
## Benchmark results (cadmium)

Specs:
* 4 x AMD Opteron(tm) Processor 6172 (2.1GHz; 48 physical cores)
* Ubuntu 16.04.6 LTS
* gcc: gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.11)
* gcc-8: gcc version 8.3.0 (Homebrew GCC 8.3.0)
* llvm: LLVM version 8.0.0

``` text
config command                                                               
C00    /home/mtf/devel/mlton/builds/g9ba73ad1e/bin/mlton -codegen amd64      
C01    /home/mtf/devel/mlton/builds/g9ba73ad1e/bin/mlton -codegen c          
C02    /home/mtf/devel/mlton/builds/g9ba73ad1e/bin/mlton -codegen c -cc gcc-8
C03    /home/mtf/devel/mlton/builds/g9ba73ad1e/bin/mlton -codegen llvm       
C04    /home/mtf/devel/mlton/builds/ga353d7851/bin/mlton -codegen amd64      
C05    /home/mtf/devel/mlton/builds/ga353d7851/bin/mlton -codegen c          
C06    /home/mtf/devel/mlton/builds/ga353d7851/bin/mlton -codegen c -cc gcc-8
C07    /home/mtf/devel/mlton/builds/ga353d7851/bin/mlton -codegen llvm       
```

### Run-Time Ratio

``` text
program           `C04/C00` `C05/C01` `C06/C02` `C07/C03`
barnes-hut           1.050     1.004     0.9939    0.9724
boyer                1.020     0.9547    0.9826    0.9940
checksum             1.007     1.034     0.9795    1.003 
count-graphs         0.9735    1.008     0.9826    0.9722
DLXSimulator         0.9397    0.9566    0.9553    0.8925
even-odd             0.9185    1.019     1.000     0.9754
fft                  1.061     1.041     1.039     1.021 
fib                  1.013     1.046     1.040     0.9884
flat-array           1.064     0.9138    0.9901    0.9247
hamlet               0.9442    0.9713    0.9664    0.9955
imp-for              1.040     0.9223    0.8787    1.005 
knuth-bendix         0.9967    1.007     0.9924    0.9744
lexgen               1.005     1.025     0.9892    0.9574
life                 0.9805    0.9813    0.9779    1.012 
logic                1.062     0.9721    1.002     0.9641
mandelbrot           0.9995    1.003     1.005     1.000 
matrix-multiply      0.9424    1.028     1.040     0.9958
md5                  0.9946    0.9913    0.9929    1.053 
merge                0.9942    0.9397    0.9797    0.9558
mlyacc               0.9956    0.9558    1.021     1.026 
model-elimination    0.9816    0.9842    1.000     0.9769
mpuz                 0.6748    0.8899    0.9354    1.084 
nucleic              0.9955    0.9901    1.008     0.9746
output1              0.9880    0.8836    0.8840    0.9090
peek                 1.010     1.075     0.9669    1.000 
pidigits             0.9778    1.055     0.9936    1.018 
psdes-random         0.9905    0.9943    1.054     1.000 
ratio-regions        1.030     1.071     0.9943    0.9448
ray                  1.016     0.9944    0.9993    1.030 
raytrace             1.024     1.009     0.9994    1.017 
simple               0.9900    1.017     1.019     1.032 
smith-normal-form    1.015     1.017     1.080     1.061 
string-concat        0.9619    1.004     0.9863    1.173 
tailfib              0.9867    1.056     1.001     0.9989
tailmerge            1.005     1.002     0.9844    0.9697
tak                  0.9886    1.025     0.9999    1.013 
tensor               1.015     0.9728    1.006     0.9927
tsp                  1.028     0.9878    0.9956    0.9745
tyan                 1.029     1.011     1.012     0.9404
vector32-concat      0.7653    0.6395    0.8281    1.003 
vector64-concat      0.8517    0.7782    0.8571    0.9619
vector-rev           0.9514    0.9847    0.8541    0.9161
vliw                 1.028     0.9368    1.017     1.011 
wc-input1            0.9170    0.9157    0.9084    0.9240
wc-scanStream        0.9929    1.007     1.146     1.038 
zebra                0.9984    1.008     1.008     0.9941
zern                 1.026     0.9929    1.005     0.9942
MIN                  0.6748    0.6395    0.8281    0.8925
GMEAN                0.9811    0.9771    0.9846    0.9912
MAX                  1.064     1.075     1.146     1.173 
```

Notes:
* Overall, there is less run-time improvements than were hoped,
  although there are some "big" wins and no "big" losses.
* The LLVM codegen seems to benefit the least; it may be that LLVM is
  able to realizes some stack-slot accesses as hardware registers
  (although it is unclear how it justifies doing so --- LLVM should
  not be able to "see" that the access of stack slots and heap objects
  do not alias).
* The speedups with `vector32-concat` and `vector64-concat` may be
  related to the observation made in MLton/mlton#288, where it was
  observed that moving a sequence from a register to a global had a
  significant slowdown (since the access through the global required
  an additional level of indirection).  In this case, a sequence may
  be moving from a stack slot to a register, eliminating a level of
  indirection.

### Compile-Time Ratio

``` text
program           `C04/C00` `C05/C01` `C06/C02` `C07/C03`
barnes-hut           0.9721    0.9760    0.9171    0.9393
boyer                0.9054    0.9675    0.9988    1.024 
checksum             0.9636    0.9630    0.9753    1.016 
count-graphs         0.9822    1.002     0.9862    0.9151
DLXSimulator         0.9205    0.9248    0.9340    0.9704
even-odd             0.9808    0.9600    0.9705    1.015 
fft                  0.9938    0.9644    0.9817    0.9656
fib                  0.9670    0.9777    0.9750    0.9473
flat-array           0.9781    0.9758    1.013     0.9805
hamlet               0.9942    1.104     1.029     1.000 
imp-for              0.9380    0.9764    0.9429    0.9979
knuth-bendix         0.9409    0.9584    0.8814    0.8520
lexgen               0.9317    0.8911    0.8933    0.9525
life                 0.9598    0.9610    0.9812    0.9046
logic                0.9704    0.9620    0.9474    0.9856
mandelbrot           0.9370    0.9926    1.011     0.9694
matrix-multiply      0.9510    0.9425    0.9634    0.9497
md5                  0.9568    0.9609    0.9473    0.9282
merge                0.9896    0.9517    0.9780    0.9987
mlyacc               1.124     1.076     1.161     1.150 
model-elimination    0.9912    0.9493    1.093     0.9879
mpuz                 0.9568    0.9690    1.006     0.9763
nucleic              0.9010    0.9169    0.9384    1.079 
output1              0.9668    0.9896    0.9598    0.9086
peek                 0.9713    0.9736    0.9840    0.9626
pidigits             0.9720    0.9686    0.9494    0.9379
psdes-random         0.9761    1.004     1.002     0.9435
ratio-regions        0.9844    0.9421    0.9679    0.9262
ray                  0.9605    0.8728    0.8533    0.8406
raytrace             0.9130    0.9501    0.8719    1.029 
simple               0.8696    0.9130    0.9537    0.9602
smith-normal-form    0.9000    0.9400    0.9632    1.167 
string-concat        0.9667    0.9934    0.9466    1.000 
tailfib              0.9807    0.9598    0.9779    0.9713
tailmerge            0.9848    0.9871    0.9878    0.9870
tak                  0.9720    1.0000    0.9728    0.9968
tensor               0.9739    0.9348    0.9057    0.8996
tsp                  0.9955    1.001     0.9283    0.9518
tyan                 0.9734    0.9536    0.9052    0.9488
vector32-concat      0.9527    0.9997    0.9673    0.9471
vector64-concat      0.9617    0.9722    0.9921    0.9885
vector-rev           0.9809    0.9846    0.9625    0.9955
vliw                 1.020     1.004     1.111     1.086 
wc-input1            0.9345    1.002     0.9502    0.9549
wc-scanStream        0.9890    0.9830    1.027     1.028 
zebra                0.9712    0.9204    1.101     0.9702
zern                 0.9416    1.013     1.098     0.8998
MIN                  0.8696    0.8728    0.8533    0.8406
GMEAN                0.9635    0.9691    0.9740    0.9727
MAX                  1.124     1.104     1.161     1.167 
```

Notes:
* One concern with assigning more RSSA variables to registers is that
  increases the burden on the codegen (hardware) register allocator.
* With the exception of `mlyacc`, there is a general improvement in
  compile time.  Note that configs C04, C05, C06, and C07 are using a
  MLton compiled with `BounceVars` to compile programs with
  `BounceVars`.  It isn't clear how much of the benefit is due to the
  effect of `BounceVars` on the compiled MLton and how much is due to
  the effect of `BounceVars` on the IR of the compiled programs.  In
  any case, it seems that `BounceVars` "pays for itself", although the
  greatest improvements are not on the largest benchmark programs
  (`hamlet`, `mlyacc`, `model-elimination`, `nucleic`, `vliw`).

### Executable-Size Ratio

``` text
program           `C04/C00` `C05/C01` `C06/C02` `C07/C03`
barnes-hut            1.005    1.005     1.009     0.9982
boyer                 1.002    0.9940    0.9967    0.9972
checksum              1.005    1.001     1.001     1.000 
count-graphs          1.007    1.004     1.005     1.001 
DLXSimulator          1.015    1.005     1.006     1.007 
even-odd              1.005    1.001     0.9999    1.0000
fft                   1.009    1.006     1.004     1.003 
fib                   1.005    1.001     0.9998    1.000 
flat-array            1.005    1.001     1.001     0.9993
hamlet                1.002    1.003     1.004     0.9999
imp-for               1.005    1.001     1.001     0.9991
knuth-bendix          1.004    0.9990    1.003     1.001 
lexgen                1.025    1.001     1.002     1.004 
life                  1.003    0.9999    1.000     0.9999
logic                 1.004    1.003     1.001     1.002 
mandelbrot            1.005    1.001     1.001     0.9994
matrix-multiply       1.005    1.000     1.002     0.9993
md5                   1.005    1.002     1.002     1.007 
merge                 1.004    1.000     0.9994    0.9995
mlyacc                1.176    1.154     1.183     1.180 
model-elimination     1.006    1.009     1.011     1.003 
mpuz                  1.006    0.9998    1.001     1.001 
nucleic               1.000    0.9996    0.9998    0.9985
output1               1.006    1.002     1.002     1.006 
peek                  1.005    1.001     1.001     0.9996
pidigits              1.003    0.9998    1.001     1.005 
psdes-random          1.005    1.001     1.002     0.9976
ratio-regions         1.020    1.008     1.007     1.009 
ray                   1.023    1.021     1.018     1.013 
raytrace              1.008    1.003     1.003     1.011 
simple                1.009    1.033     1.034     1.055 
smith-normal-form     1.005    1.005     1.006     1.003 
string-concat         1.004    0.9997    1.002     0.9980
tailfib               1.005    1.000     1.000     0.9998
tak                   1.005    1.001     0.9997    0.9999
tensor                1.019    1.019     1.018     1.020 
tsp                   1.005    1.002     1.004     1.007 
tyan                  1.034    1.026     1.025     1.030 
vector32-concat       1.005    1.000     1.001     0.9984
vector64-concat       1.005    1.000     1.001     0.9982
vector-rev            1.005    1.001     1.001     0.9988
vliw                  1.033    1.010     1.012     1.027 
wc-input1             1.012    1.009     1.008     1.010 
wc-scanStream         1.011    1.006     1.007     1.007 
zebra                 1.004    1.001     1.002     1.005 
zern                  1.006    1.003     1.001     1.002 
MIN                   1.000    0.9940    0.9967    0.9972
GMEAN                 1.012    1.007     1.008     1.008 
MAX                   1.176    1.154     1.183     1.180 
```

Notes:
* Consistent with the increase in compile-time for `mlyacc`, there is
  an increase in code size for `mlyacc`.
* Otherwise, there is little change in code size.


## Benchmark results (sulfur)

Specs:
* 2 x Intel(R) Xeon(R) CPU E5-2637 v3 @ 3.50GHz (8 physical cores; 16 logical cores)
* Ubuntu 16.04.6 LTS
* gcc: gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.11)
* gcc-8: gcc version 8.3.0 (Homebrew GCC 8.3.0)
* llvm: LLVM version 8.0.0

``` text
config command                                                               
C00    /home/mtf/devel/mlton/builds/g9ba73ad1e/bin/mlton -codegen amd64      
C01    /home/mtf/devel/mlton/builds/g9ba73ad1e/bin/mlton -codegen c          
C02    /home/mtf/devel/mlton/builds/g9ba73ad1e/bin/mlton -codegen c -cc gcc-8
C03    /home/mtf/devel/mlton/builds/g9ba73ad1e/bin/mlton -codegen llvm       
C04    /home/mtf/devel/mlton/builds/ga353d7851/bin/mlton -codegen amd64      
C05    /home/mtf/devel/mlton/builds/ga353d7851/bin/mlton -codegen c          
C06    /home/mtf/devel/mlton/builds/ga353d7851/bin/mlton -codegen c -cc gcc-8
C07    /home/mtf/devel/mlton/builds/ga353d7851/bin/mlton -codegen llvm       
```

### Run-Time Ratio

``` text
program           `C04/C00` `C05/C01` `C06/C02` `C07/C03`
barnes-hut           1.066     1.004     0.9963    0.8840
boyer                0.9890    0.9917    0.9902    0.9567
checksum             1.065     1.001     0.9935    0.9681
count-graphs         0.9580    0.9805    0.9665    0.9672
DLXSimulator         0.7266    0.7235    0.7374    0.7496
even-odd             1.009     0.9903    0.9873    0.9845
fft                  0.9930    0.9983    0.9710    0.9682
fib                  1.108     1.023     1.0000    0.9724
flat-array           0.9638    0.9581    0.9604    0.9620
hamlet               0.8991    0.9415    0.9295    0.9544
imp-for              0.9688    0.9359    0.9240    1.123 
knuth-bendix         1.005     0.9915    1.025     1.005 
lexgen               1.001     0.9954    0.9583    0.9974
life                 0.9659    1.003     0.9652    0.9923
logic                0.9887    0.9938    0.9702    0.9919
mandelbrot           0.9989    1.015     1.024     1.008 
matrix-multiply      1.017     0.9927    1.014     0.9881
md5                  0.9846    1.006     0.9868    1.026 
merge                1.071     1.077     1.039     1.074 
mlyacc               0.9994    0.9986    1.012     1.011 
model-elimination    0.9850    0.9656    0.9885    0.9879
mpuz                 0.7695    0.8649    0.9525    1.018 
nucleic              0.9267    0.9614    0.9359    0.9804
output1              0.9453    0.8869    0.8673    0.8782
peek                 0.9988    1.000     1.011     0.9938
pidigits             0.9734    0.9961    1.003     0.9986
psdes-random         0.9902    0.9998    1.016     0.9956
ratio-regions        1.107     0.9887    0.9315    0.9852
ray                  0.9905    0.9674    0.9930    0.9931
raytrace             1.008     1.032     1.005     1.057 
simple               0.9659    0.9692    0.9842    0.9648
smith-normal-form    0.9858    0.9876    0.9712    0.9615
string-concat        0.9901    1.001     1.007     0.9731
tailfib              0.9968    1.015     1.000     0.9936
tailmerge            0.9893    0.9932    0.9810    0.8081
tak                  0.9724    1.058     0.9762    0.9693
tensor               1.005     0.9982    0.9914    1.006 
tsp                  1.030     1.049     1.037     1.032 
tyan                 1.007     0.9887    0.9850    1.002 
vector32-concat      0.3328    0.3212    0.6566    1.029 
vector64-concat      0.3983    0.3806    0.7111    0.9922
vector-rev           0.9282    1.032     0.9473    0.9013
vliw                 0.9068    0.8996    0.8662    0.9419
wc-input1            1.396     0.9463    1.120     1.348 
wc-scanStream        0.9928    1.015     1.117     0.9642
zebra                0.9957    0.9966    0.9974    0.9967
zern                 0.9958    1.013     1.000     0.9743
MIN                  0.3328    0.3212    0.6566    0.7496
GMEAN                0.9468    0.9393    0.9640    0.9827
MAX                  1.396     1.077     1.120     1.348 
```

Notes:
* Again, there are some "big" wins and no "big" losses.
* Interestingly, compared to `cadmium`, `vector32-concat` and
  `vector64-concat` see greater speedups, though `mpuz` sees a smaller
  speedup.

### Compile-Time Ratio

``` text
program           `C04/C00` `C05/C01` `C06/C02` `C07/C03`
barnes-hut           0.9401    0.9632    1.013     1.006 
boyer                0.9659    1.000     0.9725    0.9629
checksum             1.022     0.9911    0.9966    0.9967
count-graphs         1.005     1.042     1.004     1.011 
DLXSimulator         0.9910    0.9526    0.9833    0.9914
even-odd             0.9654    0.9591    1.009     0.9795
fft                  0.9841    0.9713    0.9885    0.9923
fib                  0.9417    0.9946    1.000     0.9976
flat-array           1.002     1.049     0.9716    1.012 
hamlet               1.103     1.018     0.9917    0.9781
imp-for              0.9586    1.041     0.9824    1.032 
knuth-bendix         0.9690    1.007     1.034     0.9347
lexgen               1.049     0.9718    0.9930    1.029 
life                 0.9490    0.9798    1.023     1.018 
logic                0.9936    0.9303    0.9768    0.9715
mandelbrot           0.9959    0.9518    0.9339    0.9740
matrix-multiply      1.117     1.060     1.013     0.9176
md5                  1.007     0.9454    0.9904    0.9973
merge                0.9922    0.9493    0.9793    1.015 
mlyacc               1.113     1.125     1.136     1.143 
model-elimination    0.9702    1.030     1.025     1.017 
mpuz                 1.015     0.9249    1.053     0.9401
nucleic              0.9813    1.007     0.9919    1.043 
output1              0.9285    0.9746    0.9603    1.049 
peek                 0.9547    0.9915    0.9410    0.9543
pidigits             0.9569    0.9757    0.9745    0.9973
psdes-random         0.9687    0.9835    1.006     0.9596
ratio-regions        0.9640    1.021     1.019     1.036 
ray                  1.079     0.9156    0.9577    0.9714
raytrace             1.008     0.9028    1.010     1.052 
simple               1.025     1.154     0.9570    1.068 
smith-normal-form    0.9611    0.9399    0.9215    0.9196
string-concat        1.023     0.9968    0.9789    1.072 
tailfib              0.9869    0.9748    0.9711    1.001 
tailmerge            0.9812    0.9766    0.9731    0.9051
tak                  0.9694    0.9863    0.9719    0.9681
tensor               0.9759    1.061     0.9847    1.030 
tsp                  0.9818    1.094     0.9497    1.066 
tyan                 1.018     1.004     0.9927    1.011 
vector32-concat      0.9783    0.9755    0.9766    0.9792
vector64-concat      0.9820    0.9876    0.9830    0.9607
vector-rev           0.9648    0.9215    1.066     1.002 
vliw                 1.075     1.045     1.035     1.052 
wc-input1            1.012     1.025     1.005     0.9931
wc-scanStream        0.9747    0.9883    0.9940    0.9761
zebra                0.9838    0.9624    0.9758    1.003 
zern                 0.8921    0.9604    0.9794    0.9880
MIN                  0.8921    0.9028    0.9215    0.9051
GMEAN                0.9921    0.9920    0.9919    0.9985
MAX                  1.117     1.154     1.136     1.143 
```

Notes:
* Compared to `cadmium`, there is less general improvement in compile
  time, though (with the exception of `mlyacc`), no "across the board"
  slowdowns.

### Executable-Size Ratio

``` text
program           `C04/C00` `C05/C01` `C06/C02` `C07/C03`
barnes-hut            1.005    1.005     1.009     0.9983
boyer                 1.002    0.9940    0.9967    0.9972
checksum              1.005    1.001     1.001     1.000 
count-graphs          1.007    1.004     1.005     1.001 
DLXSimulator          1.015    1.005     1.006     1.008 
even-odd              1.005    1.001     0.9999    0.9999
fft                   1.009    1.006     1.004     1.003 
fib                   1.005    1.001     0.9998    1.000 
flat-array            1.005    1.001     1.001     0.9995
hamlet                1.002    1.003     1.004     0.9999
imp-for               1.005    1.001     1.001     0.9991
knuth-bendix          1.004    0.9989    1.003     1.000 
lexgen                1.025    1.001     1.002     1.004 
life                  1.003    0.9998    1.000     0.9997
logic                 1.004    1.003     1.001     0.9997
mandelbrot            1.005    1.001     1.001     0.9994
matrix-multiply       1.005    1.000     1.002     0.9996
md5                   1.005    1.002     1.002     1.007 
merge                 1.004    1.000     0.9993    0.9994
mlyacc                1.176    1.154     1.183     1.179 
model-elimination     1.006    1.009     1.011     1.003 
mpuz                  1.006    0.9998    1.001     1.001 
nucleic               1.000    0.9996    0.9998    0.9985
output1               1.006    1.002     1.002     1.006 
peek                  1.005    1.001     1.001     0.9994
pidigits              1.003    0.9998    1.001     1.005 
psdes-random          1.005    1.001     1.002     0.9978
ratio-regions         1.020    1.008     1.007     1.009 
ray                   1.023    1.021     1.018     1.013 
raytrace              1.008    1.003     1.003     1.013 
simple                1.009    1.033     1.034     1.055 
smith-normal-form     1.005    1.005     1.006     1.003 
string-concat         1.004    0.9995    1.002     0.9980
tailfib               1.005    1.000     1.000     0.9998
tailmerge             1.004    0.9995    1.001     0.9998
tak                   1.005    1.001     0.9997    1.000 
tensor                1.019    1.019     1.018     1.019 
tsp                   1.005    1.002     1.004     1.007 
tyan                  1.034    1.026     1.025     1.031 
vector32-concat       1.005    1.000     1.001     0.9984
vector64-concat       1.005    1.000     1.001     0.9982
vector-rev            1.005    1.001     1.001     0.9987
vliw                  1.033    1.010     1.012     1.026 
wc-input1             1.012    1.009     1.008     1.010 
wc-scanStream         1.011    1.006     1.007     1.007 
zebra                 1.004    1.001     1.002     1.005 
zern                  1.006    1.003     1.001     1.002 
MIN                   1.000    0.9940    0.9967    0.9972
GMEAN                 1.011    1.007     1.008     1.008 
MAX                   1.176    1.154     1.183     1.179 
```

Notes:
* As they should be (same versions of MLton, gcc, llvm), executable
  sizes are identical with `cadmium`.
